### ISSUE TITLE
In CLI output, make it clearer when operations are success or failure.

### ISSUE THREAD

**Issue by Martinsos** [wasp-lang/wasp#212]
Right now we print some messages like "Wasp generated code" or "Db operations succeeded" and similar, and there is also a lot of messages coming from client, server, db.

Sometimes, when a lot of output is produced, it can be really hard to figure out if everything together succeeded or not! That seems to be especially true for the `wasp db` commands, where output from Prisma can be confusing.

We should look into making it more obvious when operation is success or not.

Ideas;
- Use progress bars / numbers / percentages.
- Use emojis, specifically green checkmark! It is very visible and gives clear feedback. Also, red Xs for failure.
- Clearly delineate subprocesses / steps. E.g. if we have "wasp is compiling ..." and then some output and then "wasp compiled successfully", it would be great if he can somehow connect the two messages, either by indentation, number, color, or smth else. Somehow mark it as [start] ... [end], as a whole.

**Comment by shayneczyzewski** [comment]
In thinking through some of the options we have, they range from basic to pretty, and from simple to hard, of course. :)

- One of the simpler ones is showing a progress bar as it scrolls. There are libraries for this, but the challenge is the other output will cause it to scroll out of view, and many of our commands are black boxes where we don't have any insight into how far along it is, so we won't have progress to report. I decided against that one for now.

- The heavier end is something like ncurses. This would allow us to do most anything, but is also like building a GUI. :D If we did go this route, we could use a sticky footer for persistent messages/statuses, and we could even have a split pane where Wasp messages are in one scrollable area and server/db/client are in another, so they don't ever intermix. I don't have a ton of experience in this area, however, and I'm not sure what the library/platform support is. Feels like a big undertaking.

- A possible quick win is what I have been tinkering with for a day or so. What if we just tab over the server/db/client output, so visually you can see it is different, and leave Wasp messages left aligned. We could also use unicode icons to represent things like: starting (🤞 ) maybe too corny, just a placeholder; success (✅); failure (❌).

Although simple, the tab for external program output does make it easy to see Wasp messages. Thoughts on this simple direction @Martinsos @matijaSos? Does it address the main issues you had in mind here?

<img width="1002" alt="Screen Shot 2021-12-22 at 11 04 01 AM" src="https://user-images.githubusercontent.com/523636/147121612-1210effb-67c1-42cd-b203-2dc219cd7000.png">
<img width="1002" alt="Screen Shot 2021-12-22 at 11 04 09 AM" src="https://user-images.githubusercontent.com/523636/147121628-6c5d7632-176e-47b8-ba1e-4138abd7a3ca.png">
<img width="1002" alt="Screen Shot 2021-12-22 at 11 04 16 AM" src="https://user-images.githubusercontent.com/523636/147121640-5a430e8f-c8f3-4606-87d0-c1af9cd787db.png">
<img width="1002" alt="Screen Shot 2021-12-22 at 11 04 30 AM" src="https://user-images.githubusercontent.com/523636/147121663-aae3cba7-540c-4a32-b69e-ec423e19ca8c.png">

If we do go this route, I still have a few things to do:
- Add `waspSays` helpers for each of those three messages and use everywhere in CLI.
- Figure out this buffering/laziness issue where sometimes the prefixed message from Server ends up on one line, but the message ends up on the next. (see final screenshot "Compiled with warnings."). Also make the line breaks consistent.
- Make sure the unicode characters I used are cross platform/terminal friendly everywhere. Consider fallback strategy.
- Verify how error behavior is messaged when `throwError` is used.
- Force wrap long lines? (like my file path in 3rd screenshot). Would require knowing terminal width.
- ... probably a few others :D

Thanks for the early feedback, and no worries if the direction is abort lol. It's why I showed it so early. :) 👍🏻

**Comment by shayneczyzewski** [comment]
Prisma does complicate this tab idea, however, as it seems to move the cursor to overwrite lines as it writes.

**Comment by shayneczyzewski** [comment]
Stripping ANSI escape codes is kinda fun :D https://gist.github.com/shayneczyzewski/659934dcf03be5abd680c8cc8b80a9ee

**Comment by Martinsos** [comment]
Awesome analysis @shayneczyzewski !

**TLDR**; all makes sense and direction 3) sounds like a way to go for now -> only I would indent less (2 spaces) or maybe use some other method, so we don't loose much space.

So, I some time ago I put a lot of effort into making our output-forwarding logic work in such  manner that emojis and in-line-rewriting (with `\r`) work as they should, and even further that prisma commands behave interactively with no issues. That said, I did forget quite a lot about what I was doing, however I am sure I could remember quickly if needed. I will try to respond to stuff here without much preparation, but if needed I can dive deeper into specific things and remind myself.

1. I agree that progress bar won't really work hm, since we can control when the external output is coming, which pushes our output + as you said external output is black box so we can't really know the progress. Btw we could un-black box it a bit by "hardcoding" specific terms that we expect from 'node' output, or from 'prisma' output. I know that is breakable and needs maintenance, but at the end I think it won't require much maintenance and it could have a very cool effect of us understanding what is happening. But we can leave that for later.
2. ncurses -> yeah, I think that is the way to go really. It is also what we aim for in the Powerline issue. There are a couple of projects in Haskell that I know about that have nice TUI (Terminal User Interface): https://github.com/kowainik/summoner and https://gitlab.haskell.org/haskell/ghcup-hs, so it might be worth investigating how they do it. I actually think we can give this a try sooner than later, because I don't think it will be that hard to get something basic going. Ok maybe not as the very first iteration, but as the second, I would already go for it.
3. Yeah, I think this simpler approach is a way to go for the first iteration -> emojis, visual indication of what is what. We already use colors: Wasp output is yellow, while external output is prefixed with Server:, Web app: and Db: and they also each have distinct color. I thought this might be visually distinct enough from Wasp messages -> but you think we should also indent them? If so, maybe we can make it a smaller indent -> 2 spaces, or max 4? Space in terminal is short so I am not keen on wasting much of it. Maybe there is another way of visually separating Wasp messages from external ones? Maybe put extra newlines around Wasp messages? But yeah, this seems like a right direction for first try! I like the emojis you presented! And, as mentioned in the very first comment in this issue, we could also pay a bit of attention how to visually connect start and end of the same command -> if there is a way to somehow indicate that.

Hm I am surprised by this issue with text going into a new line! It seems that happens only when Wasp message gets printed? Because I know I put a lot of effort for such stuff not to happen when external input is getting mixed, but it might be that Wasp messages somehow get involved and mess things up.

Force wrapping long lines -> I wouldn't bother with it for now, terminal wraps them for us anyway.

Prisma complicating the tab approach -> but, if I am correct, the fact that Prisma does overwrite doesn't stop us from putting out `Db:` prefix in front of it, right? If so, tab is also not a problem. But I am remembering this wrong. Basically it comes down to use detecting that Prisma emitted `\r` and then we can add tab + prefix if we want, immediately after it. So it shouldn't be a big problem to do.

Ha all this is pretty fun indeed :D! Terminal UI yay :D.

**Comment by Martinsos** [comment]
Just quickly checked and `summoner` is using https://hackage.haskell.org/package/brick for its TUI -> it looks like a potentially perfect solution for what we need.

**Comment by shayneczyzewski** [comment]
Thanks @Martinsos good call on indentation, I think 2 spaces makes sense if we continue on indent path.

Thanks for the feedback on 1-3. Cool idea on (1) to unblackbox by looking for keywords, but agree it is fragile. Agree (2) is the ultimate way to go, I can do a quick review on brick to see if it isn't as big as it feels on first pass. On (3) I was thinking the indentation did slightly help visually and somewhat "connect" the commands, but can go either way if we want to indent. Emojis seem to help, I did try to for some indicator of start and end of the command with a vertical bar connecting them, may continue along that line of thought. That was just hard since some of the other output gets interspersed and it is hard to logically know if we are "in the middle of some Wasp command" as we get other command output to display, but I can try a few more things tomorrow.

Agree on not wrapping long lines even if we indent.

I did try using the DB: prefix command output for Prisma output and it went all over the place. :/ It is because of all the cursor ANSI `\ESC` escape codes Prisma uses in its output, not  just `\r`. Stuff from here: https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797 Prisma basically moves the cursor around when doing things to rewrite old lines, etc. Makes it more complicated (possibly too much so) to try and reformat nicely.

Will check out https://hackage.haskell.org/package/brick more tomorrow though, thanks for the lead and feedback!

**Comment by shayneczyzewski** [comment]
> Hm I am surprised by this issue with text going into a new line! It seems that happens only when Wasp message gets printed? Because I know I put a lot of effort for such stuff not to happen when external input is getting mixed, but it might be that Wasp messages somehow get involved and mess things up.

Yeah it is only when Wasp messages get printed, but it may not have been noticed before. But if you use master you can get it to mix Wasp and process output:

<img width="708" alt="Screen Shot 2021-12-22 at 7 57 03 PM" src="https://user-images.githubusercontent.com/523636/147171746-7b0f93ec-aa78-4e7f-a79c-7c99aeda72e2.png">

I wonder if it is when we do the adding the prefix step to the message if some portion of that is lazy, or if some output buffer needs flushing perhaps?

**Comment by Martinsos** [comment]
Output mixing -> yeah it might be that something needs to be flushed! We should look into it.

Ok, all sounds good, and I think you are right that having a vertical line connecting Wasp output is too hard for now, I would skip that.

Prisma and prefixing with DB: Hm I thought that works but ok if it doesn't! But I don't think there is any other way in terminal to rewrite the line then by using `\r`? So if we add `DB: ` after every `\r`, I don't think anything else can mess us up? But in any case, not worth spending time on that either right now.

**Comment by shayneczyzewski** [comment]
Sounds good on flushing, I can keep an eye on it.

On Prisma and DB output, agree may not be worth a ton of time but here is an example of the issue for our future selves. :) 

Current output:
<img width="1002" alt="Screen Shot 2021-12-23 at 7 07 12 AM" src="https://user-images.githubusercontent.com/523636/147238486-4ccce601-3651-4ff1-8f58-1fbb365d38d6.png">

If you swap from the simple `printJobMsgsUntilExitReceived` to the `readJobMessagesAndPrintThemPrefixed` version:
<img width="1002" alt="Screen Shot 2021-12-23 at 7 06 13 AM" src="https://user-images.githubusercontent.com/523636/147238585-36cd49e6-79c2-45ae-97a8-6cfd9c58af88.png">

You will see the difference:
<img width="1002" alt="Screen Shot 2021-12-23 at 7 06 06 AM" src="https://user-images.githubusercontent.com/523636/147238624-74beebf2-f5d3-4881-b208-bed6a703aa8f.png">

Now you see intermediate output, like the pink lines, plus all the extra prefixes, and lastly the missing start of the final output line. This is because Prisma uses additional ANSI escape codes to control the cursor and clear lines: https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#cursor-controls

I am not an expert here, but I think this means checking for `\r` and `\n` are insufficient because things like `ESC[0K` can do thinks like "erase from cursor to end of line". So if we do not strip them, when we let them pass through the Prisma output to our output, the terminal will act on them still.

**Comment by Martinsos** [comment]
> Sounds good on flushing, I can keep an eye on it.
> 
> On Prisma and DB output, agree may not be worth a ton of time but here is an example of the issue for our future selves. :)
> 
> Current output: <img alt="Screen Shot 2021-12-23 at 7 07 12 AM" width="1002" src="https://user-images.githubusercontent.com/523636/147238486-4ccce601-3651-4ff1-8f58-1fbb365d38d6.png">
> 
> If you swap from the simple `printJobMsgsUntilExitReceived` to the `readJobMessagesAndPrintThemPrefixed` version: <img alt="Screen Shot 2021-12-23 at 7 06 13 AM" width="1002" src="https://user-images.githubusercontent.com/523636/147238585-36cd49e6-79c2-45ae-97a8-6cfd9c58af88.png">
> 
> You will see the difference: <img alt="Screen Shot 2021-12-23 at 7 06 06 AM" width="1002" src="https://user-images.githubusercontent.com/523636/147238624-74beebf2-f5d3-4881-b208-bed6a703aa8f.png">
> 
> Now you see intermediate output, like the pink lines, plus all the extra prefixes, and lastly the missing start of the final output line. This is because Prisma uses additional ANSI escape codes to control the cursor and clear lines: https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#cursor-controls
> 
> I am not an expert here, but I think this means checking for `\r` and `\n` are insufficient because things like `ESC[0K` can do things like "erase from cursor to end of line". So if we do not strip them, when we let them pass through the Prisma output to our output, the terminal will act on them still.
> 
> Edit: Actually, that last paragraph and the last screenshot may not align. It seems like the escape codes are doing things in the first example, but they are not doing them in the second, but there is still some weird interaction happening in the output buffer. thinking Perhaps it is due to whatever else the prefix function is doing?

Ah wow I didn't know those cursor controls exist! For some reason I thought `\r` is the only way to control cursor to go backwards, but of course it is not, this makes a lot of sense.
